submissions/VBystrov/memory-pair-game/js/index.js

```diff
@@ -0,0 +1,166 @@
+  constructor(cardsContainer, newGameButton) {
+    this.cardsContainer = cardsContainer;
+    this.newGameButton = newGameButton;
+    this.allCardCodes = [];
+    this.pairedCards = [];
+    this.fieldLocked = false;
+
+    this.addAllCardCodes();
+
+    this.newGameButton.addEventListener('click', () => {
+      this.start();
+    });
+
+    this.cardsContainer.addEventListener('click', (e) => {
+      const card = e.target.closest('.card');
+
+      if (this.pairedCards.includes(card) || !card || this.fieldLocked) {
+        return;
+      }
+      this.reverseCards(card);
+
+      if (!(this.pairedCards.length & 1)) {
+        this.pairedCards.push(card);
+        return;
+      } else {
+        if (
+          this.isCardsPair(card, this.pairedCards[this.pairedCards.length - 1])
+        ) {
+          this.pairedCards.push(card);
+          this.pairedCards.slice(-2).forEach((newPairCard) => {
+            newPairCard.classList.add('pair');
+          });
+
+          if (this.pairedCards.length >= cardsInGame * 2) {
+            this.newGameButton.disabled = false;
+          }
+        } else {
+          this.reverseNotPairedCards(card);
+        }
+      }
+    });
+  }
+
+  start() {
+    this.newGameButton.disabled = true;
+    this.pickCards();
+    this.shuffleCards();
+    this.placeCards();
+    this.revealAllCards();
+  }
+
+  addAllCardCodes() {
+    for (let i = 0; i < amountCards; i++) {
+      this.allCardCodes.push(i.toString(10).padStart(3, '0'));
+    }
+  }
+
+  pickCards() {
+    for (let i = this.allCardCodes.length - 1; i > 0; i--) {
+      let j = Math.floor(Math.random() * (i + 1));
+      [this.allCardCodes[i], this.allCardCodes[j]] = [
+        this.allCardCodes[j],
+        this.allCardCodes[i],
+      ];
+    }
+    this.pickedCardCodes = this.allCardCodes.slice(0, cardsInGame);
+  }
+
+  shuffleCards() {
+    this.shuffledCardCodes = [];
+    for (let i = 0; i < this.pickedCardCodes.length; i++) {
+      this.shuffledCardCodes.push(
+        this.pickedCardCodes[i],
+        this.pickedCardCodes[i]
+      );
+    }
+
+    for (let i = this.shuffledCardCodes.length - 1; i > 0; i--) {
+      let j = Math.floor(Math.random() * (i + 1));
+      [this.shuffledCardCodes[i], this.shuffledCardCodes[j]] = [
+        this.shuffledCardCodes[j],
+        this.shuffledCardCodes[i],
+      ];
+    }
+  }
+
+  isCardsPair(card1, card2) {
+    return card1.attributes.cardCode.value === card2.attributes.cardCode.value;
+  }
+
+  reverseCards(...cards) {
+    cards.forEach((card) => {
+      card.classList.toggle('reverse');
+    });
+  }
+
+  reverseNotPairedCards(card) {
+    this.fieldLocked = true;
+    setTimeout(() => {
+      this.reverseCards(card, this.pairedCards.pop());
+      this.fieldLocked = false;
+    }, showCardsTime);
+  }
+
+  placeCards() {
+    this.cardsContainer.innerHTML = '';
+    for (let i = 0; i < this.shuffledCardCodes.length; i++) {
+      this.cardsContainer.innerHTML += this.createCardElement(
```

Comment:
You are accessing the DOM in a loop. DOM accesses are expensive, so it is better to batch them. Please think of a way to attach all cards to the DOM in one operation. You can create elements without attaching them to the DOM

Reply:
Done.